src/algebraic_geometry/limits.lean

```diff
@@ -0,0 +1,133 @@
+/-
+Copyright (c) 2022 Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrew Yang
+-/
+import algebraic_geometry.pullbacks
+import algebraic_geometry.AffineScheme
+import category_theory.limits.constructions.finite_products_of_binary_products
+import category_theory.limits.constructions.equalizers
+
+/-!
+# (Co)Limits of Schemes
+
+We construct various limits and colimits in the category of schemes.
+
+* The existence of fibred products was shown in `algebraic_geometry/pullbacks.lean`.
+* `Spec ℤ` is the terminal object.
+* The preceding two results imply that `Scheme` has all finite limits.
+* The empty scheme is the (strict) initial object.
+
+## Todo
+
+* Coproducts exists (and the forgetful functors preserve them).
+
+-/
+
+universe u
+
+open category_theory category_theory.limits opposite topological_space
+
+namespace algebraic_geometry
+
+/-- `Spec ℤ` is the terminal object in the category of schemes. -/
+noncomputable
+def Spec_Z_is_terminal : is_terminal (Scheme.Spec.obj (op $ CommRing.of ℤ)) :=
+@@is_terminal.is_terminal_obj _ _ Scheme.Spec _ infer_instance
+  (terminal_op_of_initial CommRing.Z_is_initial)
+
+instance : has_terminal Scheme := has_terminal_of_has_terminal_of_preserves_limit Scheme.Spec
+
+instance : is_affine (⊤_ Scheme.{u}) :=
+is_affine_of_iso (preserves_terminal.iso Scheme.Spec).inv
+
+instance : has_finite_limits Scheme :=
+has_finite_limits_of_has_terminal_and_pullbacks
+
+section initial
+
+/-- The map from the empty scheme. -/
+@[simps]
+def Scheme.empty_to (X : Scheme.{u}) : ∅ ⟶ X :=
+⟨{ base := ⟨λ x, pempty.elim x, by continuity⟩,
+    c := { app := λ U, CommRing.punit_is_terminal.from _ } }, λ x, pempty.elim x⟩
+
+@[ext]
+lemma Scheme.empty_ext {X : Scheme.{u}} (f g : ∅ ⟶ X) : f = g :=
+by { ext a, exact pempty.elim a }
+
+lemma Scheme.eq_empty_to {X : Scheme.{u}} (f : ∅ ⟶ X) : f = Scheme.empty_to X :=
+Scheme.empty_ext f (Scheme.empty_to X)
+
+instance (X : Scheme.{u}) : unique (∅ ⟶ X) :=
+⟨⟨Scheme.empty_to _⟩, λ _, Scheme.empty_ext _ _⟩
+
+/-- The empty scheme is the initial object in the category of schemes. -/
+def empty_is_initial : is_initial (∅ : Scheme.{u}) :=
```

Comment:
Could you please add this as a `def` for every scheme `Y` (generalising `∅`) that satisfies `[is_empty Y.carrier]`?

I think you can then avoid `is_open_immersion_of_is_empty` and give a short direct proof of `is_iso_of_is_empty`.

Reply:
I think `is_open_immersion_of_is_empty` is strictly more general and is worth having nonetheless?

Reply:
Yes, you're right.